### PR TITLE
Ini files

### PR DESCRIPTION
**Pull request by edersondisouza**

### CODE REVIEW

File: ini.cpp [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+{
+    FILE *file;
+    char *entry = NULL;
+    int line = 0, ret = 0;
+    size_t n = 0;
+    ssize_t read;
+
+    assert(filename);
+
+    file = fopen(filename, "r");
```
Comment (lucasdemarchi) [inline ini.cpp]: "re"

File: ini.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    if (!file) {
+        log_error_errno(errno, "Could not open ini file '%s' (%m)", filename);
```
Comment (lucasdemarchi) [inline ini.cpp]: here and everywhere: s/ini/conf/  or s/ini/config/

File: ini.cpp [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        return -EIO;
+    }
+
+    while ((read = getline(&entry, &n, file)) >= 0) {
+        line++;
+
+        // Remove trailing '/n'
```
Comment (lucasdemarchi) [inline ini.cpp]: \n

File: ini.cpp [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        if (*(entry + read - 1) == '\n')
```
Comment (lucasdemarchi) [inline ini.cpp]: what about trailing \t, spaces, etc?

File: ini.cpp [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    // just make parser add on known section
+    section = sections;
+    while (section) {
+        if (!strcmp(section->name, entry)) {
+            current_section = section;
+            free(entry);
+            return 0;
+        }
+        section = section->next;
+    }
```
Comment (lucasdemarchi) [inline ini.cpp]: a method find_section()?

File: ini.cpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+    // Section is new.
+    section = (struct section *)calloc(1, sizeof(struct section));
+    null_check(section, -ENOMEM);
```
Comment (lucasdemarchi) [inline ini.cpp]: ugh... looks like my comments were not addressed when this null_check() got in

File: ini.cpp [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        log_error("On file %s: Line %d: Missing value on config", filename, line);
+        return -EINVAL;
+    }
+
+    config = (struct config *)malloc(sizeof(struct config));
+    null_check(config, -ENOMEM);
+
+    config->key = entry;
+    config->value = equal_pos + 1;
+    *equal_pos = '\0';
```
Comment (lucasdemarchi) [inline ini.cpp]: it's very common to have
Bla = "bli"

in this case you will have an extra space in the key.

File: ini.cpp [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+    config->next = current_section->configs;
+    current_section->configs = config;
+
+    return 0;
+}
+
+const char *IniFile::nextFromSection(const char *section_name, const char *key)
```
Comment (lucasdemarchi) [inline ini.cpp]: camel case?

File: ini.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+#pragma once
+
+class IniFile {
+public:
+    IniFile(const char *_filename)
+        : filename{_filename}
+        , sections{nullptr}
+        , current_section{nullptr} {};
```
Comment (lucasdemarchi) [inline ini.h]: use c++11 initialization.

File: ini.h [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    int parse_file();
+    const char *nextFromSection(const char *section_name, const char *key);
+
+private:
+    const char *filename;
+    struct section *sections;
+    struct section *current_section;
+
+    int add_section(char *entry, int line);
+    int add_config(char *entry, int line);
```
Comment (lucasdemarchi) [inline ini.h]: missing underscores on private members

File: README.md [hunk 11]
```diff
@@ -43,18 +43,42 @@ Install:
...
+`MAVLINK_ROUTER_INI_FILE` environment variable, or via `-i` switch when running
+mavlink-routerd.
+An example of ini file would be:
+
+```
+[General]
+tcp=5790
+report-stats=false
+
+[Master-Endpoints]
```
Comment (lucasdemarchi) [inline README.md]: should be endpoint only. "master" is a property of the endpoint, that can be renamed to eavesdrop

File: README.md [hunk 12]
```diff
@@ -43,18 +43,42 @@ Install:
...
+UART=/dev/tty0:52000
+UDP=0.0.0.0:10000
+UDP=0.0.0.0:10001
+
+[Endpoints]
+UDP=127.0.0.1:11000
+UDP=127.0.0.1:11001
+```
+
```
Comment (lucasdemarchi) [inline README.md]: should be:

```
[ Endpoint A]
Type = UART
Options = baud=460800
Mode = Eavesdropping

[ Endpoint B]
Type = UART
Options = baud=115200
Mode = Normal

[ Endpoint C]
Type = UDP
Options = port=11000
Mode = Eavesdropping

```

File: ini.cpp (hunk 4 above)
Comment (edersondisouza) [inline ini.cpp]: Fair.

File: ini.cpp (hunk 6 above)
Comment (edersondisouza) [inline ini.cpp]: Hmmm... what were the comments? Anyway, I can drop using it here.

File: ini.cpp (hunk 7 above)
Comment (edersondisouza) [inline ini.cpp]: `trim()` everywhere.

File: ini.cpp (hunk 8 above)
Comment (edersondisouza) [inline ini.cpp]: ops!

File: ini.cpp (hunk 2 above)
Comment (lucasdemarchi) [inline ini.cpp]: to be clear: extension should be .conf.

File: ini.h (hunk 9 above)
Comment (edersondisouza) [inline ini.h]: Err... isn't this already c++11 initialization? What am I missing?

File: README.md (hunk 12 above)
Comment (edersondisouza) [inline README.md]: In this example, 'A' would be endpoint address, like '0.0.0.0' or '/dev/tty0', or this information would appear on yet another key=value?

File: ini.h (hunk 9 above)
Comment (lucasdemarchi) [inline ini.h]: no. This is:

```
class Bla {
    int bli = 42;
    const char *bla = "42";
};
```

Use the initialization list only for stuff that is going to change when the arguments change

File: README.md (hunk 12 above)
Comment (edersondisouza) [inline README.md]: And not sure if this suboption `Options = port=11000` is anything beautiful. Why not, then something like:

```ini
[Endpoint blah]
Type = UDP
Port = 1000

[Endpoint foo]
Type = uart
Baud = 115200
```

?

Comment (lucasdemarchi) [inline README.md]: I was thinking about a key that would be generic for any endpoint and still allow custom options. But I think it's ok the way you wrote it

**Comment by edersondisouza** [comment]
Updated to address @lucasdemarchi comments:
 - Conf file instead of ini file
 - Using `_` on non-public members
 - Format of file changed

File: conf.cpp [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+
+    while ((read = getline(&entry, &n, file)) >= 0) {
+        line++;
+
+        _trim(&entry);
+
+        switch (entry[0]) {
+        case ';':
+        case '\0':
+            // Discards comment or blank line
```
Comment (lucasdemarchi) [inline conf.cpp]: '#' is more common on Linux for commenting ini files. ';' is so windows

**Comment by edersondisouza** [comment]
v2 on https://github.com/01org/mavlink-router/pull/20
(Create a new version as there was kinda big addition - drop in overriding for configurations)

**Pull request closed without merging**